File: recursion/reverseString.js
var reverseString = function(s) {
    const N = s.length;
    
    for(let i = 0; i < N / 2; i++) {
        const j = N - 1 - i;
        [s[i], s[j]] = [s[j], s[i]];
    }
};

var reverseString2 = function(s) {
    start = 0 ;
    end = s.length-1;
    return helper(s,start,end);
};

function helper(s,start,end){
    if (start<end){
        temp = s[start];
        s[start] = s[end];
        s[end] = temp;
        return helper(s,start+1,end-1);
    } else{
        return s;
    }
};

File: recursion/reverseString.py
class Solution:
    def reverseString(self, s):
        """
        Do not return anything, modify s in-place instead.
        input: list
        output: reversed list
        two pointers: one from 0 position, another one from the end
        tmp: store end temporialy 
        """
        start,end = 0,len(s)-1
        def helper(start,end):
            if start < end:
                s[start],s[end] = s[end],s[start]
                helper(start+1,end-1)
        helper(start,end)
        return s 

s = Solution()
print(s.reverseString(["a","b","c"]))
